### ISSUE TITLE
Parsing fails in last RUN heredoc when no subsequent newlines at end of Dockerfile

### ISSUE THREAD

**Issue by arledesma** [hadolint/hadolint#736]
- [x] This is a bug report
- [ ] This is a feature request
- [x] I searched existing issues before opening this one

### Expected behavior

Expecting `RUN` HEREDOC to be parsed correctly when end of heredoc is the last line of the Dockerfile with no additional newlines following the `RUN` with the HEREDOC.

Similar error to #699 

### Actual behavior

Parsing fails only when there is not an additional newline at the end of the Dockerfile.

```shell
╭─
╰─❯ docker run --rm -i hadolint/hadolint hadolint - < <(<Dockerfile)
/dev/stdin:6:1 unexpected 'e' expecting '#', '\', ADD, ARG, CMD, COPY, ENTRYPOINT, ENV, EXPOSE, FROM, HEALTHCHECK, LABEL, MAINTAINER, ONBUILD, RUN, SHELL, STOPSIGNAL, USER, VOLUME, WORKDIR, a pragma, at least one space, or end of input

╭─
╰─❯ sed -n '6p' Dockerfile
echo "GOODBYE WORLD" > /tmp/goodbye
```

Adding the additional newline at the end of the Dockerfile then succeeds.

```shell
╭─
╰─❯ docker run --rm -i hadolint/hadolint hadolint - < <(<Dockerfile; echo;)
-:4 DL3002 warning: Last USER should not be root
╭─
╰─❯ 
```

### Steps to reproduce the behavior

**Output of `hadolint --version` or
  `docker run --rm hadolint/hadolint hadolint --version` or
  `docker run --rm ghcr.io/hadolint/hadolint hadolint --version`:**

```shell
╭─ 
╰─❯ docker run --rm hadolint/hadolint hadolint --version
Haskell Dockerfile Linter v2.8.0-0-g398770f-dirty
```

<details>
<summary><code>docker build</code></summary>

```bash
╭─ 
╰─❯ docker build -t hadolint-bug .
[+] Building 24.7s (12/12) FINISHED
 => [internal] load build definition from Dockerfile                                                                                 0.0s
 => => transferring dockerfile: 156B                                                                                                 0.0s
 => [internal] load .dockerignore                                                                                                    0.0s
 => => transferring context: 2B                                                                                                      0.0s
 => resolve image config for docker.io/docker/dockerfile:1.3-labs                                                                   12.0s
 => [auth] docker/dockerfile:pull token for registry-1.docker.io                                                                     0.0s
 => CACHED docker-image://docker.io/docker/dockerfile:1.3-labs@sha256:250ce669e1aeeb5ffb892b18039c3f0801466536cb4210c8eb2638e628859  0.0s
 => [internal] load build definition from Dockerfile                                                                                 0.0s
 => [internal] load .dockerignore                                                                                                    0.0s
 => [internal] load metadata for docker.io/library/alpine:3.15                                                                      11.9s
 => [auth] library/alpine:pull token for registry-1.docker.io                                                                        0.0s
 => CACHED [1/2] FROM docker.io/library/alpine:3.15@sha256:fb150366bfb5a297a7f8852e0cec462a12f638374f8a04ad235c56a97e780add          0.0s
 => [2/2] RUN <<eot (echo "GOODBYE WORLD" > /tmp/goodbye)                                                                            0.4s
 => exporting to image                                                                                                               0.1s
 => => exporting layers                                                                                                              0.0s
 => => writing image sha256:8364dc293df78a8f18b8eea998f4101d521ed5722a5f4df033077768b9938234                                         0.0s
 => => naming to docker.io/library/hadolint-bug                                                                                      0.0s
```
</details>

<details>
<summary><code>docker run</code></summary>

```bash
╭─ 
╰─❯ docker run --rm -it hadolint-bug cat /tmp/goodbye
GOODBYE WORLD
```
</details>

**Dockerfile** (if relevant)

```Dockerfile
# syntax=docker/dockerfile:1.3-labs

FROM alpine:3.15
USER root
RUN <<eot
echo "GOODBYE WORLD" > /tmp/goodbye
eot
```

**Additional environment details (OS, stack version, etc.)**

Using `buildx` with `# syntax=docker/dockerfile:1.3-labs`

**Comment by m-ildefons** [comment]
Hi @arledesma 
Thanks for the bug report, this is indeed a bug in the parser library. I'd like to point out, that according to [the spec](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_07_04) the newline is not optional, but I agree that in case the file just ends this should not crash the parser.
